django/test/testcases.py

```diff
@@ -889,18 +908,23 @@ def assertInHTML(self, needle, haystack, count=None, msg_prefix=""):
         haystack = assert_and_parse_html(
             self, haystack, None, "Second argument is not valid HTML:"
         )
+        if msg_prefix:
+            msg_prefix += ": "
         real_count = haystack.count(needle)
         if count is not None:
             self.assertEqual(
                 real_count,
                 count,
-                msg_prefix
-                + "Found %d instances of '%s' in response (expected %d)"
-                % (real_count, needle, count),
+                msg_prefix + "Found %d instances of '%s' (expected %d)"
+                " in the following response\n%s"
+                % (real_count, needle, count, safe_repr(haystack)),
```

Comment:
Here and below, noting one of my comments from the ticket hasn’t been addressed:

> Additionally for assertInHTML – it’s annoying that needle as displayed in the failure message is the parser’s output, which is therefore broken up over multiple lines. It’d be much nicer if the example above resulted in AssertionError: False is not true : Couldn't find '<b>hey</b>' in response.

This is definitely in the "nice to have" territory, but I can’t think of a better place to address this than now.

To illustrate, with a `self.assertInHTML("<b>hey</b>", "<p>Howdy!</p>")`, Django’s existing output is:

```
AssertionError: False is not true : Couldn't find '<b>
hey
</b>' in response
```

The output proposed here would be:

```
AssertionError: False is not true : Couldn't find '<b>
hey
</b>' in the following response
'<p>
Howdy!
</p>'
```

My preference would be:

```
AssertionError: False is not true : Couldn't find '<b>hey</b>' in the following response
'<p>Howdy!</p>'
```

---

For this to work it’s just a matter of not overriding the `needle` and `haystack` variables at the top of the method, so we print the parameters as-is.